### ISSUE TITLE
`-e REALTIME` sometimes takes no samples

### ISSUE THREAD

**Issue by blue42u** [HPCToolkit/hpctoolkit#562]
Logging since this is still an issue in `master`.

Test case:
```console
$ cat foo.c
#include <stdio.h>
#include <time.h>
#include <errno.h>
#include <error.h>

int main() {
  struct timespec ts = {1, 0};
  errno = 0;
  int reps = 0;
  while(clock_nanosleep(CLOCK_MONOTONIC, 0, &ts, &ts) != 0 && errno == EINTR) reps++;
  if(errno != 0 && errno != EINTR)
    error(1, errno, "clock_nanosleep failed");
  printf("interrupts: %d\n", reps);
  return 0;
}
$ gcc -o foo foo.c
$ HPCRUN_LOG_STDERR=1 ./bin/hpcrun -e REALTIME -dd ITIMER_HANDLER -dd ITIMER_CTL ./foo
fnbounds: .../libexec/hpctoolkit/hpcfnbounds2
[610105, N]: ITIMER_CTL: init
[610105, N]: ITIMER_CTL: process event list, lush_metrics = 0
[610105, N]: ITIMER_CTL: checking event spec = REALTIME
[610105, N]: ITIMER_CTL: init REALTIME sample_period = 5000, seconds = 0, usec = 5000
[610105, N]: ITIMER_CTL: setting metric timer period = 1
[610105, N]: ITIMER_CTL: start REALTIME
[610105, N]: ITIMER_HANDLER: starting REALTIME: value = (0,5000), interval = (0,0)
interrupts: 0
[610105, N]: ITIMER_CTL: stop REALTIME
[610105, N]: ITIMER_CTL: stop REALTIME
[610105, N]: ITIMER_CTL: shutdown REALTIME
MEMORY: segment size: 4.0 meg, num segments: 1, total allocation: 4.0 meg, reclaims: 0
MEMORY: total freeable: 0.0 meg, total non-freeable: 0.1 meg, malloc failures: 0
UNWIND ANOMALIES: total: 0 errant: 0, total-frames: 0, total-libunwind-fails: 0
ACC SUMMARY:
         accelerator trace records: 0 (processed: 0, dropped: 0)
         accelerator samples: 0 (recorded: 0, dropped: 0)
SAMPLE ANOMALIES: blocks: 0 (async: 0, dlopen: 0), errors: 0 (segv: 0, soft: 0)
SUMMARY: samples: 0 (recorded: 0, blocked: 0, errant: 0, trolled: 0, yielded: 0),
         frames: 0 (trolled: 0)
         intervals: 0 (suspicious: 0)
```

**Comment by mwkrentel** [comment]
I'll look at this sometime after the current release,
unless someone wants to grab it first.

**Comment by jmellorcrummey** [comment]
We should craft a test case for clock_nanosleep() that highlights the failure to return EINTR and either send it to libc-alpha or hand it to Ben Woodard for action by the glibc team.

We can wait on that until we see Ben at the end of the month.

We should override these and handle the restart ourselves. We've seen this issue with hpctoolkit speeding up programs by aborting sleeps in the past.

**Comment by jmellorcrummey** [comment]
@mwkrentel please file a git issue for overriding the sleeps so we fix this eventually. At present, we have many higher priority things to work on.

**Comment by blue42u** [comment]
@jmellorcrummey Hold the press a bit, there's no Glibc bug, I misread the manual. `clock_nanosleep` returns its error code instead of setting `errno`. Should've read more carefully (although it doesn't actually make a difference for the 0-sample bug).

Updated test code:
```c
#include <stdio.h>
#include <time.h>
#include <errno.h>
#include <error.h>

int main() {
  struct timespec ts = {1, 0};
  int reps = 0;
  int err;
  while((err = clock_nanosleep(CLOCK_MONOTONIC, 0, &ts, &ts)) == EINTR) reps++;
  if(err != 0)
    error(1, err, "clock_nanosleep failed");
  printf("interrupts: %d\n", reps);
  return 0;
}
```

* * *
@mwkrentel After some more extensive testing, I've discovered that this 0-sample bug only appears with Glibc 2.33 or later, which would explain why you're having trouble reproducing it. An `ubuntu:22.04` or `fedora:36` container will have Glibc 2.35, presumably the bug will appear there.

**Comment by mwkrentel** [comment]
Yes, indeed, I also read the man page too quickly.  The `pthread_`
functions work this way, but it's been a long time since I've
programmed pthreads.

Ok, back to square one, must be something different.  But if the
latest glibc behaves differently from current glibc, it sounds like
one of them has a bug.

Do we have a machine with the latest glibc?
Or, where's a good place to run containers?  And how?

But note, the issue of overriding `usleep()`, etc still exists.
If an app is not expecting to handle signals, then our interrupts
will distort the behavior of their sleeps.  So, it makes sense
that we should correct for that.

**Comment by mwkrentel** [comment]
@jmellorcrummey @blue42u 
I've tried this in a container and I can reproduce the problem, but
I'm also getting strange results.  This is my first non-trivial
container, so I could easily be doing something wrong.

This is on ufront.cs with a docker/podman image for fedora:35, glibc
2.34-34.fc35.x86_64, plus rpm/yum packages for gcc (11.3.1), gfortran,
git, perl, python, cmake, makeinfo.

Remember, containers are inherently out of sync on one point: I'm
running fedora:35 with a RH 8.x kernel 4.18.0-348.23.1.el8_5.x86_64.
This is not the kernel that fedora 35 comes with.  That's probably not
a problem, but signals go through the kernel, so theoretically, there
could be something out of sync.

1. Yes, in the container with fedora 35 and glibc 2.34, I get no
interrupts for your revised unit test with clock_nanosleep() and
REALTIME.

Actually, I'm not sure if there should be interrupts or not.
clock_nanosleep() is a syscall, we use SA_RESTART, so if the kernel is
obeying SA_RESTART, then no interrupts should be correct.  Yes, that
would be a change in semantics somewhere, but I don't see why it's
necessarily incorrect.

----------

Some other strange results:

2. The hpcrun log files contain:

```
[236422, N]: !! INTERNAL ERROR, not possible to map dso for (null) (0x7fffe1fa7000, 0x7fffe1fa7f5d)
```

This happens with REALTIME and CPUTIME.
Do you also get this?

3. The above (2) happens even with sample source NONE.  This is
actually two bugs: one that some dso fails, and two, why are we even
trying to map a dso for NONE.

4. PAPI overflow totally fails inside the container.  This fails in
hpcrun with an error like:

```
HPCToolkit fatal error: PAPI event PAPI_TOT_CYC is not supported on this platform.
```

This is with a spack built papi.  And it also happens with my papi
unit tests outside of hpctoolkit.

So, either containers simply can't handle PAPI (that would be strange
and unfortunate), or perhaps this is a case where the kernel out of
sync with /usr actually matters.

And no, the problem is not `perf_event_paranoid`.
Inside the container `perf_event_paranoid` is 1, same as outside
the container (probably because it's the same kernel).

**Comment by mwkrentel** [comment]
Ah yes, once again I'm guilty of failing to RTFM.
So, then it's a clear bug.

Indeed, man 7 signal agrees:

```
The following interfaces are never restarted after being interrupted
by a signal handler, regardless of the use of SA_RESTART; they always
fail with the error EINTR when interrupted by a signal handler:
...
  * Sleep interfaces: clock_nanosleep(2), nanosleep(2), and usleep(3).
```

But it's still theoretically possible that a native Fedora 35 system
and a container for fedora:35 might behave differently, depending on
the underlying kernel.

**Comment by blue42u** [comment]
Fixed (2) in 1dec0052cb5b641ecd86d498a2a22ceef5052b18, it was a bug in the auditor (I forgot to supply the path to the saved vDSO on one of the logic paths).

**Comment by mwkrentel** [comment]
In `dlopen()`, if filename is NULL, then you're asking for the
handle of the main binary.  So, maybe that's what's happening
in (3).

**Comment by mwkrentel** [comment]
The problem with (4) above was that I didn't use `--cap-add=cap_sys_admin` when running podman. 
With `cap_sys_admin` I now get the same results inside the container as outside for PAPI overflow and perf.

**Comment by mwkrentel** [comment]
Well, this keeps getting stranger and stranger and harder.

I was able to reproduce this with Fedora 35 (in a container).
So, I wrote a unit test to demonstrate a problem in glibc.
Works fine (get EINTR) in the unit test.

So, works in the unit test, fails in hpcrun, gotta be a bug in hpcrun,
right?  So, why does the version of glibc matter ??

I guess it's a bug in hpcrun, or at least a bad interaction between
hpcrun and glibc.  Or perhaps an issue with the container.  Maybe the
container needs some extra capability to replicate the native
semantics.

I was hoping Jonathon's fix for INTERNAL ERROR would have fixed this
as well, but alas no.

I tried going back in hpctoolkit to 2021.03.01.  Same behavior.  So,
if it's a regression in hpcrun, it came in before 2021.03.01.  Before
then, I started running into issues building older revs.
